trees/cart_classifier.go

```diff
@@ -0,0 +1,481 @@
+package trees
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/sjwhitworth/golearn/base"
+)
+
+// The "c" prefix to function names indicates that they were tailored for classification
+
+// CNode is Node struct for Decision Tree Classifier
+type classifierNode struct {
+	Left       *classifierNode
+	Right      *classifierNode
+	Threshold  float64
+	Feature    int64
+	LeftLabel  int64
+	RightLabel int64
+	Use_not    bool
+	maxDepth   int64
+}
+
+// CARTDecisionTreeClassifier: Tree struct for Decision Tree Classifier
+type CARTDecisionTreeClassifier struct {
+	RootNode    *classifierNode
+	criterion   string
+	maxDepth    int64
+	labels      []int64
+	triedSplits [][]float64
+}
+
+// Calculate Gini Impurity of Target Labels
+func giniImpurity(y []int64, labels []int64) (float64, int64) {
+	nInstances := len(y)
+	gini := 0.0
+	maxLabelCount := 0
+	var maxLabel int64 = 0
+	for label := range labels {
+		numLabel := 0
+		for target := range y {
+			if y[target] == labels[label] {
+				numLabel++
+			}
+		}
+		p := float64(numLabel) / float64(nInstances)
+		gini += p * (1 - p)
```

Comment:
Computing the gini impurity given a frequency distribution
 1 - sigma_i (p_i ^ 2)
is being done by computeGini Index method in gini.go.
Can't it be reused. 
Also the name of the methods giniImpurity, entropy is misleading if it is doing more than just computing gini Impurity of a distribution.